README.md

```diff
@@ -1,3 +1,14 @@
 # newsletter-list-cleanse
 
-A monthly job to clean newsletter mailing lists of lapsed subscribers.
\ No newline at end of file
+A monthly job to clean newsletter mailing lists of lapsed subscribers.
+
+## Structure ##
+
+The job is split into three lambdas that are joined together using SQS.
```

Comment:
Maybe this use case too simple for it, but did you consider using a step function to coordinate?

Reply:
I don't know step functions very well as I've never used them, but I think both the volume and general logic won't apply very well.

Here in each step of the way there's more and more work.
So getCutOffDates fetches a list of newsletter -> dates. Each item of that list is then independently passed to getCleanseList. Then getCleanseList take each one of these and finds the users to remove from the newsletter and batches that in little packets of 1K users for the last processing stage. Then each batch of 1K users will be updates in Braze. So the number of invocations looks like an inverted funnel, where each step is bigger than the previous one.

I might be wrong but I would see step functions more applied to cases where you have a single object and need it to transition from one state to another. I'm sure we could use step functions though, it boils down to how we model our project.

I'm happy to have a chat about it if you think it would be a better fit?

Reply:
I believe step functions do support "fan-out" use-cases like this nowadays, you can even control things like the level of parallelism.

My instinct would be that it's quite simple to just use lambda/SQS for this initially if there are only 3 lambdas, but if you start to break down those stages any further, or introduce new ones then the balance might tip the other way. In my experience step function's main benefit is ease of maintenance through better debugging/replay/monitoring - so if your use case is unlikely to be high maintenance then it may not be worth it.

Reply:
Ha, that's very good to know, I'll have a look then as this might help us wait for athena queries to come back!